In [1]:
#필요한 import
import pandas as pd
import numpy as np

## 1. movie 데이터의 크기를 줄여봅시다. 

In [2]:
# movie데이터 원본
movie = pd.read_csv('data/movie.csv')
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      4897 non-null   object 
 1   director_name              4814 non-null   object 
 2   num_critic_for_reviews     4867 non-null   float64
 3   duration                   4901 non-null   float64
 4   director_facebook_likes    4814 non-null   float64
 5   actor_3_facebook_likes     4893 non-null   float64
 6   actor_2_name               4903 non-null   object 
 7   actor_1_facebook_likes     4909 non-null   float64
 8   gross                      4054 non-null   float64
 9   genres                     4916 non-null   object 
 10  actor_1_name               4909 non-null   object 
 11  movie_title                4916 non-null   object 
 12  num_voted_users            4916 non-null   int64  
 13  cast_total_facebook_likes  4916 non-null   int64

In [3]:
# 11개의 column 선택과 결측값 채우기
movie1 = movie.iloc[:, :11].fillna(0)

In [4]:
movie1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   color                    4916 non-null   object 
 1   director_name            4916 non-null   object 
 2   num_critic_for_reviews   4916 non-null   float64
 3   duration                 4916 non-null   float64
 4   director_facebook_likes  4916 non-null   float64
 5   actor_3_facebook_likes   4916 non-null   float64
 6   actor_2_name             4916 non-null   object 
 7   actor_1_facebook_likes   4916 non-null   float64
 8   gross                    4916 non-null   float64
 9   genres                   4916 non-null   object 
 10  actor_1_name             4916 non-null   object 
dtypes: float64(6), object(5)
memory usage: 422.6+ KB


In [5]:
movie1.describe().T

count          mean           std  min        25%  \
num_critic_for_reviews   4916.0  1.366135e+02  1.204214e+02  0.0      47.00   
duration                 4916.0  1.067640e+02  2.592923e+01  0.0      93.00   
director_facebook_likes  4916.0  6.766770e+02  2.805134e+03  0.0       6.00   
actor_3_facebook_likes   4916.0  6.283228e+02  1.622638e+03  0.0     129.00   
actor_1_facebook_likes   4916.0  6.485241e+03  1.509821e+04  0.0     606.00   
gross                    4916.0  3.929025e+07  6.380675e+07  0.0  285641.25   

                                50%          75%          max  
num_critic_for_reviews        106.5       191.00        813.0  
duration                      103.0       118.00        511.0  
director_facebook_likes        44.0       188.00      23000.0  
actor_3_facebook_likes        362.0       631.00      23000.0  
actor_1_facebook_likes        981.5     11000.00     640000.0  
gross                    14914785.0  50875351.25  760505847.0

In [6]:
# memory_usage를 통해 메모리 사용량을 column별로 알 수 있다.
movie1.memory_usage(deep=True)

Index                         128
color                      306262
director_name              343388
num_critic_for_reviews      39328
duration                    39328
director_facebook_likes     39328
actor_3_facebook_likes      39328
actor_2_name               346635
actor_1_facebook_likes      39328
gross                       39328
genres                     379929
actor_1_name               346599
dtype: int64

In [7]:
np.finfo('float16')

finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)

In [8]:
np.finfo('float32')

finfo(resolution=1e-06, min=-3.4028235e+38, max=3.4028235e+38, dtype=float32)

In [9]:
np.finfo('float64')

finfo(resolution=1e-15, min=-1.7976931348623157e+308, max=1.7976931348623157e+308, dtype=float64)

In [10]:
movie1.memory_usage(deep=True)

Index                         128
color                      306262
director_name              343388
num_critic_for_reviews      39328
duration                    39328
director_facebook_likes     39328
actor_3_facebook_likes      39328
actor_2_name               346635
actor_1_facebook_likes      39328
gross                       39328
genres                     379929
actor_1_name               346599
dtype: int64

In [11]:
#최대값이 813임으로 가장 작은 float16이 들어간다.
movie1['num_critic_for_reviews'] = movie1['num_critic_for_reviews'].astype(np.float16)
movie1.memory_usage(deep=True)

Index                         128
color                      306262
director_name              343388
num_critic_for_reviews       9832
duration                    39328
director_facebook_likes     39328
actor_3_facebook_likes      39328
actor_2_name               346635
actor_1_facebook_likes      39328
gross                       39328
genres                     379929
actor_1_name               346599
dtype: int64

In [12]:
print(9832 / 39328 * 100, '%')

25.0 %


원래 데이터의 크기에 비해 25%로 줄어들었다.

In [13]:
#최대값이 511임으로 가장 작은 float16이 들어간다.
movie1['duration'] = movie1['duration'].astype(np.float16)
movie1.memory_usage(deep=True)

Index                         128
color                      306262
director_name              343388
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes     39328
actor_3_facebook_likes      39328
actor_2_name               346635
actor_1_facebook_likes      39328
gross                       39328
genres                     379929
actor_1_name               346599
dtype: int64

In [14]:
print(9832 / 39328 * 100, '%')

25.0 %


원래 데이터의 크기에 비해 25%로 줄어들었다.

In [15]:
#최대값이 23000임으로 가장 작은 float16이 들어간다.
movie1['director_facebook_likes'] = movie1['director_facebook_likes'].astype(np.float16)
movie1.memory_usage(deep=True)

Index                         128
color                      306262
director_name              343388
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes      39328
actor_2_name               346635
actor_1_facebook_likes      39328
gross                       39328
genres                     379929
actor_1_name               346599
dtype: int64

In [16]:
print(9832 / 39328 * 100, '%')

25.0 %


원래 데이터의 크기에 비해 25%로 줄어들었다.

In [17]:
#최대값이 23000임으로 가장 작은 float16이 들어간다.
movie1['actor_3_facebook_likes'] = movie1['actor_3_facebook_likes'].astype(np.float16)
movie1.memory_usage(deep=True)

Index                         128
color                      306262
director_name              343388
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes       9832
actor_2_name               346635
actor_1_facebook_likes      39328
gross                       39328
genres                     379929
actor_1_name               346599
dtype: int64

In [18]:
print(9832 / 39328 * 100, '%')

25.0 %


원래 데이터의 크기에 비해 25%로 줄어들었다.

In [19]:
#최대값이 640000임으로 float32이 들어간다.
movie1['actor_1_facebook_likes'] = movie1['actor_1_facebook_likes'].astype(np.float32)
movie1.memory_usage(deep=True)

Index                         128
color                      306262
director_name              343388
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes       9832
actor_2_name               346635
actor_1_facebook_likes      19664
gross                       39328
genres                     379929
actor_1_name               346599
dtype: int64

In [20]:
print(19664 / 39328 * 100, '%')

50.0 %


원래 데이터의 크기에 비해 50%로 줄어들었다.

In [21]:
#최대값이 760505847임으로 float32이 들어간다.
movie1['gross'] = movie1['gross'].astype(np.float32)
movie1.memory_usage(deep=True)

Index                         128
color                      306262
director_name              343388
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes       9832
actor_2_name               346635
actor_1_facebook_likes      19664
gross                       19664
genres                     379929
actor_1_name               346599
dtype: int64

In [22]:
print(19664 / 39328 * 100, '%')

50.0 %


원래 데이터의 크기에 비해 50%로 줄어들었다.

In [23]:
movie1.describe(include=['object']).T

count unique             top  freq
color          4916      3           Color  4693
director_name  4916   2398               0   102
actor_2_name   4916   3031  Morgan Freeman    18
genres         4916    914           Drama   233
actor_1_name   4916   2096  Robert De Niro    48

In [24]:
movie1.select_dtypes(include=['object']).nunique()

color               3
director_name    2398
actor_2_name     3031
genres            914
actor_1_name     2096
dtype: int64

위의 두 개의 함수를 통해 object들의 unique값을 알 수 있다.

총 데이터의 수와 unique의 값이 비슷하면 category로 변환하는 것이 큰 의미가 없다.

그렇기 때문에 count가 4916인데 비해 작은 color(3) 정도가 유효할 것이라고 생각하고 진행해 보았다.

In [25]:
movie1['color'] = movie1['color'].astype('category')
movie1.memory_usage(deep=True)

Index                         128
color                        5190
director_name              343388
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes       9832
actor_2_name               346635
actor_1_facebook_likes      19664
gross                       19664
genres                     379929
actor_1_name               346599
dtype: int64

In [26]:
print(5190 / 306262 * 100, '%')

1.6946274758213555 %


원래 데이터 크기에 비해 1.69%로 줄어들었다.

이렇게 데이터 타입을 바꾸어 데이터 손실 없이 데이터를 줄일 수 있다.

In [27]:
movie1['genres'] = movie1['genres'].astype('category')
movie1.memory_usage(deep=True)

Index                         128
color                        5190
director_name              343388
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes       9832
actor_2_name               346635
actor_1_facebook_likes      19664
gross                       19664
genres                     121102
actor_1_name               346599
dtype: int64

In [28]:
print(121102 / 379929 * 100, '%')

31.87490294239187 %


In [29]:
movie1['director_name'] = movie1['director_name'].astype('category')
movie1.memory_usage(deep=True)

Index                         128
color                        5190
director_name              245814
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes       9832
actor_2_name               346635
actor_1_facebook_likes      19664
gross                       19664
genres                     121102
actor_1_name               346599
dtype: int64

In [30]:
print(245814 / 343388 * 100, '%')

71.58491269351288 %


In [31]:
movie1['actor_2_name'] = movie1['actor_2_name'].astype('category')
movie1.memory_usage(deep=True)

Index                         128
color                        5190
director_name              245814
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes       9832
actor_2_name               290652
actor_1_facebook_likes      19664
gross                       19664
genres                     121102
actor_1_name               346599
dtype: int64

In [32]:
print(290652 / 346635 * 100, '%')

83.84958241377818 %


In [33]:
movie1['actor_1_name'] = movie1['actor_1_name'].astype('category')
movie1.memory_usage(deep=True)

Index                         128
color                        5190
director_name              245814
num_critic_for_reviews       9832
duration                     9832
director_facebook_likes      9832
actor_3_facebook_likes       9832
actor_2_name               290652
actor_1_facebook_likes      19664
gross                       19664
genres                     121102
actor_1_name               224147
dtype: int64

In [34]:
print(224147 / 346599 * 100, '%')

64.67041162842364 %


앞서 설명한 것과 같이 color를 제외한 genres, director_name, actor_2_name, actor_1_name는 unique값이 너무 커 category화 하여도 데이터가 늘어날 것이라 생각했는데 결과적으로는 줄어들었다.

## 2. college 에서 UGDS 의 최대값을 구하는데 도시별로.
> INSTNM, CITY, UGDS

In [35]:
#college데이터
college = pd.read_csv('data/college.csv')
college.head()

INSTNM        CITY STABBR  HBCU  MENONLY  \
0             Alabama A & M University      Normal     AL   1.0      0.0   
1  University of Alabama at Birmingham  Birmingham     AL   0.0      0.0   
2                   Amridge University  Montgomery     AL   0.0      0.0   
3  University of Alabama in Huntsville  Huntsville     AL   0.0      0.0   
4             Alabama State University  Montgomery     AL   1.0      0.0   

   WOMENONLY  RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY  ...  UGDS_2MOR  \
0        0.0         0     424.0     420.0           0.0  ...     0.0000   
1        0.0         0     570.0     565.0           0.0  ...     0.0368   
2        0.0         1       NaN       NaN           1.0  ...     0.0000   
3        0.0         0     595.0     590.0           0.0  ...     0.0172   
4        0.0         0     425.0     430.0           0.0  ...     0.0098   

   UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  PCTPELL  PCTFLOAN  UG25ABV  \
0    0.0059     0.0138    0.0656         1   0.7356    0.8284   0.1049   
1    0.0179     0.0100    0.2607         1   0.3460    0.5214   0.2422   
2    0.0000     0.2715    0.4536         1   0.6801    0.7795   0.8540   
3    0.0332     0.0350    0.2146         1   0.3072    0.4596   0.2640   
4    0.0243     0.0137    0.0892         1   0.7347    0.7554   0.1270   

   MD_EARN_WNE_P10  GRAD_DEBT_MDN_SUPP  
0            30300               33888  
1            39700             21941.5  
2            40100               23370  
3            45500               24097  
4            26600             33118.5  

[5 rows x 27 columns]

In [36]:
#sort_values를 사용한 방법
(
    college[['INSTNM', 'CITY', 'UGDS']]
    .sort_values('UGDS', ascending=False)
)

INSTNM              CITY  \
7116                      University of Phoenix-Arizona             Tempe   
1189                         Ivy Tech Community College      Indianapolis   
793                                  Miami Dade College             Miami   
3711                           Lone Star College System     The Woodlands   
3669                          Houston Community College           Houston   
...                                                 ...               ...   
7530         SAE Institute of Technology  San Francisco        Emeryville   
7531                  Rasmussen College - Overland Park     Overland Park   
7532  National Personal Training Institute of Cleveland  Highland Heights   
7533  Bay Area Medical Academy - San Jose Satellite ...          San Jose   
7534            Excel Learning Center-San Antonio South       San Antonio   

          UGDS  
7116  151558.0  
1189   77657.0  
793    61470.0  
3711   59920.0  
3669   58084.0  
...        ...  
7530       NaN  
7531       NaN  
7532       NaN  
7533       NaN  
7534       NaN  

[7535 rows x 3 columns]

INSTNM, CITY, UGDS 3가지 컬럼에 대해서 UGDS의 최대값을 기준으로 정렬했다.

In [37]:
(
    college[['INSTNM', 'CITY', 'UGDS']]
    .sort_values('UGDS', ascending=False)
    .drop_duplicates(subset='CITY')
)

INSTNM                CITY  \
7116                 University of Phoenix-Arizona               Tempe   
1189                    Ivy Tech Community College        Indianapolis   
793                             Miami Dade College               Miami   
3711                      Lone Star College System       The Woodlands   
3669                     Houston Community College             Houston   
...                                            ...                 ...   
7469          Strayer University-Warrendale Campus          Warrendale   
7479          Rasmussen College-Mokena/Tinley Park              Mokena   
7485  Strayer University-Palm Beach Gardens Campus  Palm Beach Gardens   
7486       Strayer University-Coral Springs Campus       Coral Springs   
7504         Strayer University-Willingboro Campus         Willingboro   

          UGDS  
7116  151558.0  
1189   77657.0  
793    61470.0  
3711   59920.0  
3669   58084.0  
...        ...  
7469       NaN  
7479       NaN  
7485       NaN  
7486       NaN  
7504       NaN  

[2514 rows x 3 columns]

.drop_duplicates(subset='CITY')을 통해 CITY를 기준으로 중복을 지워주어 CITY를 그룹으로 만들어 CITY별로 최대값을 만들었다.

In [38]:
(
    college[['INSTNM', 'CITY', 'UGDS']]
    .sort_values(['CITY', 'UGDS'], ascending=False)
    .drop_duplicates(subset='CITY')
)

INSTNM          CITY    UGDS
2916                                 Zane State College    Zanesville  2063.0
83                              Arizona Western College          Yuma  7218.0
4833                           Yukon Beauty College Inc         Yukon    25.0
6561                           Elite Cosmetology School  Yucca Valley    91.0
296                               Crafton Hills College       Yucaipa  5152.0
...                                                 ...           ...     ...
3307  Pennsylvania State University-Penn State Abington      Abington  3505.0
3926               Virginia Highlands Community College      Abingdon  1703.0
3610                       Abilene Christian University       Abilene  3572.0
3492                          Northern State University      Aberdeen  1693.0
6241                                  Angeles Institute       ARTESIA   114.0

[2514 rows x 3 columns]

정렬을 위해 sort_values에 CITY를 추가했다.

In [39]:
#groupby를 사용한 방법
(
    college[['INSTNM', 'CITY', 'UGDS']]
    .groupby('CITY', as_index=False)
    .apply(lambda df: df.sort_values(
    'UGDS', ascending=False).head(1)
    )
.droplevel(0)
.sort_values('CITY', ascending=False)
)

INSTNM          CITY    UGDS
2916                                 Zane State College    Zanesville  2063.0
83                              Arizona Western College          Yuma  7218.0
4833                           Yukon Beauty College Inc         Yukon    25.0
6561                           Elite Cosmetology School  Yucca Valley    91.0
296                               Crafton Hills College       Yucaipa  5152.0
...                                                 ...           ...     ...
3307  Pennsylvania State University-Penn State Abington      Abington  3505.0
3926               Virginia Highlands Community College      Abingdon  1703.0
3610                       Abilene Christian University       Abilene  3572.0
3492                          Northern State University      Aberdeen  1693.0
6241                                  Angeles Institute       ARTESIA   114.0

[2514 rows x 3 columns]

groupby를 사용한 방법으로 INSTNM, CITY, UGDS 3가지 컬럼에서 

.groupby('CITY', as_index=False)를 통해 CITY를 기준으로 그룹을 잡고

.apply(lambda df: df.sort_values('UGDS', ascending=False).head(1))를 통해 그 그룹들 중에서 한 개를 선택한다. 그리고 UGDS로 정렬을 한다.(내림차순)

.droplevel(0).sort_values('CITY', ascending=False)를 통해 앞의 인덱스를 지워준다.

In [40]:
#nlargest를 사용한 방법
(
    college[['INSTNM', 'CITY', 'UGDS']]
    .nlargest(10000, 'UGDS')
    .drop_duplicates(subset='CITY')
    #.sort_values('CITY', ascending=False)
)

INSTNM           CITY  \
7116                      University of Phoenix-Arizona          Tempe   
1189                         Ivy Tech Community College   Indianapolis   
793                                  Miami Dade College          Miami   
3711                           Lone Star College System  The Woodlands   
3669                          Houston Community College        Houston   
...                                                 ...            ...   
5316   Marlboro College Graduate & Professional Studies    Brattleboro   
5957                   Gentle Healing School of Massage       Cranbury   
1674                                     Hebrew College  Newton Centre   
1889  Hazelden Betty Ford Graduate School of Addicti...    Center City   
654                   Lyme Academy College of Fine Arts       Old Lyme   

          UGDS  
7116  151558.0  
1189   77657.0  
793    61470.0  
3711   59920.0  
3669   58084.0  
...        ...  
5316       7.0  
5957       5.0  
1674       2.0  
1889       1.0  
654        0.0  

[2446 rows x 3 columns]

college[['INSTNM', 'CITY', 'UGDS']]를 통해 3가지 컬럼을 가져왔다.

.nlargest(10000, 'UGDS')를 통해 UGDS의 값이 가장 큰 10000개의 데이터를 가져왔다.

.drop_duplicates(subset='CITY')를 통해 CITY를 그룹으로 주어 CITY별로 UGDS의 최대값을 가져왔다.

nlargest를 사용할 때는 NA값이 제외되어 rows값이 다르게 나온다.

## 3. 주별, RELAFFIL 별 UGDS의 최대값을 구하라
> INSTNM, STABBR, RELAFFIL, UGDS

In [41]:
#sort_values를 사용한 방법
(
college[['INSTNM', 'STABBR', 'RELAFFIL', 'UGDS']]
.sort_values(['STABBR', 'RELAFFIL', 'UGDS'], ascending=[True, True, False])
.drop_duplicates(subset=['STABBR', 'RELAFFIL'])
)

INSTNM STABBR  RELAFFIL     UGDS
60       University of Alaska Anchorage     AK         0  12865.0
64            Alaska Pacific University     AK         1    275.0
5             The University of Alabama     AL         0  29851.0
46                   Samford University     AL         1   3033.0
137              University of Arkansas     AR         0  21405.0
...                                 ...    ...       ...      ...
4123    University of Wisconsin-Madison     WI         0  29302.0
4085               Marquette University     WI         1   8212.0
5817  American Public University System     WV         0  44924.0
4054     West Virginia Wesleyan College     WV         1   1375.0
4137              University of Wyoming     WY         0   9910.0

[112 rows x 4 columns]

sort_values를 사용한 방법으로 INSTNM, STABBR, RELAFFIL, UGDS 컬럼에서 

.sort_values(['STABBR', 'RELAFFIL', 'UGDS'], ascending=[True, True, False])을 통해 STABBR별 RELAFFIL별에서 UGDS의 ascending을 False로 줘서 최대값으로 정렬하였다.

.drop_duplicates(subset=['STABBR', 'RELAFFIL'])을 통해 STABBR와 RELAFFIL 그룹으로 만들었다.

In [42]:
#groupby를 사용한 방법
(
    college[['INSTNM', 'STABBR', 'RELAFFIL', 'UGDS']]
    .groupby(['STABBR', 'RELAFFIL'], as_index=False)
    .apply(lambda df: df.sort_values(
    'UGDS', ascending=False).head(1)
    )
.droplevel(0)
#.sort_values(['STABBR', 'RELAFFIL', 'UGDS'], ascending=[False, True, False])
)

INSTNM STABBR  RELAFFIL     UGDS
60       University of Alaska Anchorage     AK         0  12865.0
64            Alaska Pacific University     AK         1    275.0
5             The University of Alabama     AL         0  29851.0
46                   Samford University     AL         1   3033.0
137              University of Arkansas     AR         0  21405.0
...                                 ...    ...       ...      ...
4123    University of Wisconsin-Madison     WI         0  29302.0
4085               Marquette University     WI         1   8212.0
5817  American Public University System     WV         0  44924.0
4054     West Virginia Wesleyan College     WV         1   1375.0
4137              University of Wyoming     WY         0   9910.0

[112 rows x 4 columns]

groupby를 사용한 방법으로 .groupby(['STABBR', 'RELAFFIL'], as_index=False)을 통해 STABBR와 RELAFFIL 그룹으로 만들고

.apply(lambda df: df.sort_values('UGDS', ascending=False).head(1))를 통해 그 그룹별로 한 개를 선택한다. 그리고 UGDS로 정렬을 한다.

sort_values를 사용한 방법과 같은 결과가 나온 것을 알 수 있다. 

In [43]:
#nlargest를 사용한 방법
(
    college[['INSTNM', 'STABBR', 'RELAFFIL', 'UGDS']]
    .nlargest(10000, 'UGDS')
    .drop_duplicates(subset=['STABBR', 'RELAFFIL'])
    #.sort_values(['STABBR', 'RELAFFIL', 'UGDS'], ascending=[True, True, False])
)

INSTNM STABBR  RELAFFIL      UGDS
7116    University of Phoenix-Arizona     AZ         0  151558.0
1189       Ivy Tech Community College     IN         0   77657.0
793                Miami Dade College     FL         0   61470.0
3711         Lone Star College System     TX         0   59920.0
3880               Liberty University     VA         1   49340.0
...                               ...    ...       ...       ...
4561  College of the Marshall Islands     MH         0    1078.0
4215          Palau Community College     PW         0     602.0
64          Alaska Pacific University     AK         1     275.0
5289       Pacific Islands University     GU         1      65.0
6439          Touro University Nevada     NV         1      31.0

[110 rows x 4 columns]

.nlargest(10000, 'UGDS')를 통해 UGDS의 값이 가장 큰 10000개의 데이터를 가져왔다.

.drop_duplicates(subset=['STABBR', 'RELAFFIL'])를 통해 각 그룹으로 만들어 그룹별로 UGDS의 최대값을 가져왔다.

nlargest를 사용할 때는 NA값이 제외되어 rows값이 다르게 나온다.

rows수가 groupby와 sort_values를 사용한 방법에 비해서 2개 적은데 nlargest를 사용할 때는 NA값이 제외되기 때문에 110rows가 나오게 된다.

In [44]:
(
    college[['INSTNM', 'STABBR', 'RELAFFIL', 'UGDS']]
    .nlargest(10000, 'UGDS')
    .drop_duplicates(subset=['STABBR', 'RELAFFIL'])
    .sort_values(['STABBR', 'RELAFFIL', 'UGDS'], ascending=[True, True, False])
)

INSTNM STABBR  RELAFFIL     UGDS
60       University of Alaska Anchorage     AK         0  12865.0
64            Alaska Pacific University     AK         1    275.0
5             The University of Alabama     AL         0  29851.0
46                   Samford University     AL         1   3033.0
137              University of Arkansas     AR         0  21405.0
...                                 ...    ...       ...      ...
4123    University of Wisconsin-Madison     WI         0  29302.0
4085               Marquette University     WI         1   8212.0
5817  American Public University System     WV         0  44924.0
4054     West Virginia Wesleyan College     WV         1   1375.0
4137              University of Wyoming     WY         0   9910.0

[110 rows x 4 columns]

.sort_values(['STABBR', 'RELAFFIL', 'UGDS'], ascending=[True, True, False])를 추가하면 groupby와 sort_values를 사용했을 때와 마찬가지로 정렬한다. 

# 4. 데이터를 찾기 시작해보자.

현재 조를 확정하지 못했기 때문에 조원들과 같이 데이터를 찾아보지는 못했습니다. 그렇기 때문에 일단 임의로 프로그래밍3 프로젝트를 위한 분석용 분석 대상 데이터를 찾아 보았습니다.

우선 두 가지의 데이터를 찾아보았는데 첫 번째로 movielens의 데이터를 선택했고 

그 중 MovieLens 1M Dataset 데이터를 찾아보았습니다. 이 데이터는 MovieLens 1M Dataset은 6,040명의 MovieLens 사용자가 약 3,900편의 영화에 대한 1,000,209개의 평가 데이터가 포함되어있습니다.

이 데이터는 userid, movieid, rating, timestamp로 이루어진 ratings 파일과 userid, gender, age, occupation, zip-code로 이루어진 user 파일

그리고 movieid, title, genres로 이루어진 movies 파일 이렇게 3가지의 파일로 이루어있습니다. 

grouping과 관련해서는 Gender(성별)는 2개, age(나이)는 7개, Occupation(직업)은 21개, Genres(장르)는 18개로 나누어져있습니다.

MovieLens 1M Dataset 데이터는 https://grouplens.org/datasets/movielens/ 에서 다운로드가 가능합니다.

이 데이터의 분석을 통해 얻을 수 있는 목표는 성별, 직업, 나이 등의 차이에 따른 평점을 분석하여 선호하는 영화의 장르가 무엇인지와 영화 선호도의 분석을 할 수 있을 것입니다.

두 번째로는 경찰청_범죄자 생활정도, 혼인관계 및 부모관계 데이터를 찾아보았습니다.
데이터는 범죄대분류, 범죄 중분류로 나누어지고 생활정도(4), 혼인관계(4), 미혼자부모관계(9), 미상(1)의 column으로 나누어집니다.

각 범죄자의 범죄 유형 분류와 가족관계, 생활정도에 대한 데이터 입니다.

경찰청_범죄자 생활정도, 혼인관계 및 부모관계 데이터는 공공데이터포털 https://www.data.go.kr/data/3074470/fileData.do 에서 다운로드 받을 수 있습니다.

이 데이터의 분석을 통한 목표는 범죄자의 가족관계, 생활 정도에 따라 어떤 범죄 유형이 나타나고 빈도수에 대한 분석을 할 수 있을 것입니다.